platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/crypto/EnhancedKeyStoreLoader.java

```diff
@@ -1165,4 +1175,340 @@ private interface AddressBookCallback {
+     *     <li>all *.pfx files moved to <b>OLD_PFX_KEYS</b> sub-directory and no longer used.</li>
+     *     <li>all agreement key material is deleted from disk.</li>
+     * </ul>
+     *
+     * @return this {@link EnhancedKeyStoreLoader} instance.
+     */
+    @NonNull
+    public EnhancedKeyStoreLoader migrate() throws KeyLoadingException, KeyStoreException {
+        logger.info(STARTUP.getMarker(), "Starting key store migration");
+        final Map<NodeId, PrivateKey> pfxPrivateKeys = new HashMap<>();
+        final Map<NodeId, Certificate> pfxCertificates = new HashMap<>();
+
+        // delete agreement keys permanently.  They are being created at startup by generateIfNecessary() after scan().
+        deleteAgreementKeys();
+
+        // create PEM files for signing keys and certs.
+        long errorCount = extractPrivateKeysAndCertsFromPfxFiles(pfxPrivateKeys, pfxCertificates);
+
+        if (errorCount == 0) {
+            // validate only when there are no errors extracting pem files.
+            errorCount = validateKeysAndCertsAreLoadableFromPemFiles(pfxPrivateKeys, pfxCertificates);
+        }
+
+        if (errorCount > 0) {
+            // roll back due to errors.
+            // this deletes any pem files created, but leaves the agreement keys deleted.
+            logger.error(STARTUP.getMarker(), "Due to {} errors, reverting pem file creation.", errorCount);
+            rollBackSigningKeysAndCertsChanges(pfxPrivateKeys, pfxCertificates);
+        } else {
+            // cleanup pfx files by moving them to sub-directory
+            cleanupMovePfxFilesToSubDirectory();
+        }
+
+        logger.info(STARTUP.getMarker(), "Completed migration of key store.");
+        return this;
+    }
+
+    /**
+     * Delete any agreement keys from the key store directory.
+     */
+    private void deleteAgreementKeys() {
+        // delete any agreement keys of the form a-*
+        final File[] agreementKeyFiles = keyStoreDirectory.toFile().listFiles((dir, name) -> name.startsWith("a-"));
+        if (agreementKeyFiles != null) {
+            for (File agreementKeyFile : agreementKeyFiles) {
+                if (agreementKeyFile.isFile()) {
+                    try {
+                        Files.delete(agreementKeyFile.toPath());
+                        logger.debug(STARTUP.getMarker(), "Deleted agreement key file {}", agreementKeyFile.getName());
+                    } catch (IOException e) {
+                        logger.error(
+                                ERROR.getMarker(),
+                                "Failed to delete agreement key file {}",
+                                agreementKeyFile.getName());
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Extracts the private keys and certificates from the PFX files and writes them to PEM files.
+     *
+     * @param pfxPrivateKeys  the map of extracted private keys.
```

Comment:
The javadocs for `pfxPrivateKeys` and `pfxCertificates` are a bit misleading, since the implication based on the wording is that these maps *already* contain extracted material. When, in reality, these are empty maps that will be populated by this method.

Reply:
updated wording